docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+Term L, M, N    := + (let BK B.. M)
+
+BiningKind BK   := (nonrec)
+                   (rec)
+Binding B       := (termbind VD L)
+                   (typebind TVD T)
+                   (datatypebind D)
+
+Datatype D      := (datatype
+                      TVD   # the type variable for the datatype
+                      TVD.. # the argument type variables of the datatype
+                      x     # the name of the destructor function
+                      C..   # the constructors of the datatype
+                   )
+Constructor C   := (constr
+                       x   # the name of the constructor
+                       T.. # the argument types of the constructor
+                    )
+```
+
+Here is an example, constructing `[1]`:
+```
+# bind List
+(let (rec)
+    (datatypebind
+        (datatype
+            (tyvardecl List (fun (type) (type)))
+            (tyvardecl a (type))
+            match_List
+            (constr
+                Nil
+            )
+            (constr
+                Cons
+                a
+                [List a]
+            )
+        )
+    )
+    # bind int64 for convenience
+    (typebind (tyvardecl int (type)) [(con integer) (con 64)])
+    # apply cons to 1 and nil
+    [ { Cons int } (con 64 ! 1) { Nil int } ]
+)
+```
+
+Another, implementing `fromMaybe`:
+```
+# bind Maybe
+(let (nonrec)
+    (datatypebind
+        (datatype
+            (tyvardecl Maybe (fun (type) (type)))
+            (tyvardecl a (type))
+            match_Maybe
+            (constr
+                Nothing
+            )
+            (constr
+                Just
+                a
+            )
+        )
+    )
+    (abs a (type)
+        (lam default a
+            (lam arg [Maybe a]
+                # match on the arg, producing an a, giving the default
+                # in the Nothing case and using the identity function
+                # in the Just case
+                [ { match_Maybe a } arg default (lam x a x) ]
+            )
+        )
+    )
+)
+```
+
+
+## Validity
+
+### Mutually recursive types
+
+The syntax of Plutus IR suggests that you can write mutually recursive datatypes, even though
+we don't know how to compile those yet. I think we will eventually be able to cover all
+types defined with this syntax, but until then, validity of Plutus IR programs will be somewhat
+determined simply by *whatever we are able to compile*.
+
+### Typechecking
+
+Similarly, while we will define a typing judgement for Plutus IR, we will probably not
+write a typechecker for it, and rather will compile it to Plutus Core before typechecking.
+
+### Value restriction
+
+Since type-lets are going to be desugared as type abstractions, a type-let whose body is not
+a value will fail the Plutus Core value restriction, and there is no attempt to prevent this
+in the Plutus IR syntax. It will simply compile into an invalid Plutus Core program.
+
+## Typing
+
+Typing rules are largely the same as Plutus Core.
```

Comment:
Do we need them? I think we can just hand-wave the translation to PLC and say "see the type checker of PLC". Having rules would implicitly imply that we actually check types (for me at least).

Reply:
I think it would be great (and feasible) to prove that the translation from Plutus IR to Plutus Core is type-preserving, and we can't do that unless we have a typing judgement for Plutus IR!

Reply:
Agreed.

Reply:
Yes, but that is something for later.